dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/InstanceAddressURL.java

```diff
@@ -343,37 +312,91 @@ public boolean hasMethodParameter(String method, String key) {
             return false;
         }
         return hasServiceMethodParameter(protocolServiceKey, method);
     }
+    
+    private String getParameterOnConsumerUrl(final ParameterOnConsumerUrl parameterOnConsumerUrl, final URL consumerUrl,
+                                             final String service, final String method, final String key) {
+        switch(parameterOnConsumerUrl) {
+            case Parameter:
+                return consumerUrl.getParameter(key);
+            case ServiceParameter:
+                return consumerUrl.getServiceParameter(service, key);
+            case ServiceMethodParameter:
+                return consumerUrl.getServiceMethodParameter(service, method, key);
+            case MethodParameter:
+                return consumerUrl.getMethodParameter(method, key);
+            default:
+                return null;
+        }
+    }
+    
+    private Optional<String> getParameterOnConsumerUrl(final ParameterOnConsumerUrl parameterOnConsumerUrl, final String service,
+                                                       final String method, final String key) {
+        if (!consumerParamFirst(key)) {
+            return Optional.empty();
+        }
+        URL consumerUrl = RpcContext.getServiceContext().getConsumerUrl();
+        if (consumerUrl == null) {
+            return Optional.empty();
+        }
+        String value = getParameterOnConsumerUrl(parameterOnConsumerUrl, consumerUrl, service, method, key);
+        if (StringUtils.isEmpty(value)) {
+            return Optional.empty();
+        }
+        return Optional.ofNullable(value);
+    }
+    
+    private Boolean hasParameterOnConsumerUrl(final ParameterOnConsumerUrl parameterOnConsumerUrl, final String service,
+                                              final String method, final String key, final Boolean checkConsumerParamFirst) {
+        if (checkConsumerParamFirst && !consumerParamFirst(key)) {
+            return false;
+        }
+        URL consumerUrl = RpcContext.getServiceContext().getConsumerUrl();
+        if (consumerUrl == null) {
+            return false;
+        }
+        return hasParameterOnConsumerUrl(parameterOnConsumerUrl, consumerUrl, service, method, key);
+    }
+    
+    private Boolean hasParameterOnConsumerUrl(final ParameterOnConsumerUrl parameterOnConsumerUrl, final URL consumerUrl,
+                                              final String service, final String method, final String key) {
+        switch(parameterOnConsumerUrl) {
+            case ServiceMethodParameter:
+                return consumerUrl.hasServiceMethodParameter(service, method, key);
+            case MethodParameter:
+                return consumerUrl.hasMethodParameter(method, key);
+            default:
+                return false;
+        }
+    }
+    
+    enum ParameterOnConsumerUrl {
+        Parameter,
+        ServiceParameter,
+        ServiceMethodParameter,
+        MethodParameter
+    }
```

Comment:
Will this make code more unreadable

Reply:
just replaced `consumerUrl` operation with `Function<URL, String>`

Reply:
As `using Function to refactor the switch` way?

Reply:
```
    public Optional<String> executeWithConsumerUrl (Function<URL, String> function) {
            URL consumerUrl = RpcContext.getServiceContext().getConsumerUrl();
            if (consumerUrl != null) {
                String v = function.apply(consumerUrl);
                if (StringUtils.isNotEmpty(v)) {
                    return Optional.of(v);
                }
                return Optional.empty();
            }
    }
```